test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -557,33 +588,59 @@ public synchronized void ensureAtMostNumDataNodes(int n) throws IOException {
         int ord = nextNodeId.getAndIncrement();
-        return buildNode(ord, random.nextLong(), settings, version);
+        return buildNode(ord, random.nextLong(), settings, version, false);
     }
 
     private NodeAndClient buildNode() {
         int ord = nextNodeId.getAndIncrement();
-        return buildNode(ord, random.nextLong(), null, Version.CURRENT);
+        return buildNode(ord, random.nextLong(), null, Version.CURRENT, false);
     }
 
-    private NodeAndClient buildNode(int nodeId, long seed, Settings settings, Version version) {
+    private NodeAndClient buildNode(int nodeId, long seed, Settings settings, Version version, boolean reuseExisting) {
         assert Thread.holdsLock(this);
         ensureOpen();
         settings = getSettings(nodeId, seed, settings);
         Collection<Class<? extends Plugin>> plugins = getPlugins();
-        String name = buildNodeName(nodeId);
-        assert !nodes.containsKey(name);
+        String name = buildNodeName(nodeId, settings);
+        if (reuseExisting && nodes.containsKey(name)) {
+            return nodes.get(name);
+        } else {
+            assert reuseExisting == true || nodes.containsKey(name) == false :
+                "node name [" + name + "] already exists but not allowed to use it";
+        }
         Settings finalSettings = Settings.builder()
-                .put(Environment.PATH_HOME_SETTING.getKey(), baseDir) // allow overriding path.home
-                .put(settings)
-                .put("node.name", name)
-                .put(DiscoveryNodeService.NODE_ID_SEED_SETTING.getKey(), seed)
-                .build();
+            .put(Environment.PATH_HOME_SETTING.getKey(), baseDir) // allow overriding path.home
+            .put(settings)
+            .put("node.name", name)
+            .put(DiscoveryNodeService.NODE_ID_SEED_SETTING.getKey(), seed)
+            .build();
         MockNode node = new MockNode(finalSettings, version, plugins);
         return new NodeAndClient(name, node);
     }
 
-    private String buildNodeName(int id) {
-        return nodePrefix + id;
+    private String buildNodeName(int id, Settings settings) {
+        String prefix = nodePrefix;
+        prefix = prefix + getRoleSuffix(settings);
+        return prefix + id;
+    }
+
+    /**
+     * returns a suffix string based on the node role. If no explicit role is defined, the suffix will be empty
+     */
+    private String getRoleSuffix(Settings settings) {
+        String suffix = "";
+        if (Node.NODE_MASTER_SETTING.exists(settings) && Node.NODE_MASTER_SETTING.get(settings)) {
+            suffix = suffix + DiscoveryNode.Role.MASTER.getAbbreviation();
+        }
+        if (Node.NODE_DATA_SETTING.exists(settings) && Node.NODE_DATA_SETTING.get(settings)) {
+            suffix = suffix + DiscoveryNode.Role.DATA.getAbbreviation();
+        }
+        if (Node.NODE_MASTER_SETTING.exists(settings) && Node.NODE_MASTER_SETTING.get(settings) == false &&
```

Comment:
`if (DiscoveryNode.isDataNode(settings)).` doesn't check for the existence of the setting, so an empty setting will return the default, which is true. So passing an empty mapping to this function will result in nodes being named `node_smd1` as both master and data roles will return true.